### PR TITLE
fix: correctly split quoted args

### PR DESCRIPTION
**Pull request by gregmagolan**
Fixes https://github.com/aspect-build/bazel-lib/issues/891 (requested by Bazel Open Source Support customer).

---

### Changes are visible to end-users: yes

- Searched for relevant documentation and updated as needed: yes
- Breaking change (forces users to change their own code or config): no
- Suggested release notes appear below: yes

New function `split_args` is added to the public API of `strings.bzl`.

This fix touches a number of different rules:
- `jq` rule expansions now include make vars and rule fixed to single quote args that contain spaces
- `expand_template` optimized to not to bother with pointless `.split(" ")` followed by `" ".join()`
- `bats_test` optimized to not to bother with pointless `.split(" ")` followed by `" ".join()`
- `params_file` rule fixed to correctly split args on quotes
- `run_binary` rule fixed to correctly split args on quotes 

### Test plan

- Covered by existing test cases
- New test cases added

Unit tests added for the new `split_args` function that handles parsing args and splitting them taking into account single quotes, double quotes and escapes.

### CODE REVIEW

File: lib/private/expand_locations.bzl [hunk 1]
```diff
@@ -32,6 +32,9 @@ def expand_locations(ctx, input, targets = []):
 
     Returns:
       The expanded path or the original path
+
+    Deprecated:
```
Comment (jbedard) [inline lib/private/expand_locations.bzl:36]: Why was this deprecated? Maybe put that in it's own PR instead of hiding it under "correctly split quoted args"?

File: lib/tests/base64_tests.bzl [hunk 2]
```diff
@@ -45,5 +46,5 @@ base64_test = unittest.make(_base64_test_impl)
 def base64_test_suite():
     unittest.suite(
         "base64_tests",
-        base64_test,
+        partial.make(base64_test, timeout = "short"),
```
Comment (jbedard) [inline lib/tests/base64_tests.bzl:49]: What were all these for?

Comment (gregmagolan) [reply lib/tests/base64_tests.bzl:49]: Resolves warnings. We have the flag set in this repo that warns if tests are not sized correctly.

File: lib/private/expand_locations.bzl (hunk 1 above)
Comment (gregmagolan) [reply lib/private/expand_locations.bzl:36]: Mentioned it in the PR summary. Its a pointless since it just calls straight through to `ctx.expand_location` and returns the result. In the past their used to be an implementation and it was dropped a while ago.

I can break it out into a separate PR.

File: lib/private/run_binary.bzl [hunk 3]
```diff
@@ -47,15 +47,11 @@ Possible fixes:
             rule_kind = str(ctx.attr.tool.label),
         ))
 
-    # `expand_locations(...).split(" ")` is a work-around https://github.com/bazelbuild/bazel/issues/10309
-    # _expand_locations returns an array of args to support $(execpaths) expansions.
-    # TODO: If the string has intentional spaces or if one or more of the expanded file
-    # locations has a space in the name, we will incorrectly split it into multiple arguments
     for a in ctx.attr.args:
-        args.add_all([expand_variables(ctx, e, outs = outputs) for e in expand_locations(ctx, a, ctx.attr.srcs).split(" ")])
+        args.add_all(split_args(expand_variables(ctx, ctx.expand_location(a, targets = ctx.attr.srcs), outs = outputs)))
```
Comment (jbedard) [inline lib/private/run_binary.bzl:51]: Can we split only if `expand_location` expanded to more then one file? Do we want to be splitting every arg 100% of the time? What if an arg is intentionally `foo bar` and we just want `args.add("foo bar")` to quote it for us?

Comment (jbedard) [reply lib/private/run_binary.bzl:51]: That might be a bigger question for a major semver release though? :/

File: lib/private/strings.bzl [hunk 4]
```diff
@@ -583,3 +583,73 @@ def hex(number):
         hex_string = "0"
 
     return "{}0x{}".format("-" if is_signed else "", hex_string)
+
+def split_args(s):
```
Comment (jbedard) [inline lib/private/strings.bzl:587]: Does performance matter here? I'm trying to think of how frequently this would be invoked for the typical use case. Can you think of common cases where this would end up being invoked a very frequently?

File: lib/private/run_binary.bzl (hunk 3 above)
Comment (gregmagolan) [reply lib/private/run_binary.bzl:51]: I thought about this and I think the answer is that if it expands to multiple files then it is split _unless_ you quote it in the args: 

```
args = ["$(location //multiple_files_target)"]
```
will expand to multiple args the same as
```
args = ["--one --two --three"]
```
would expand to multiple args

but you can prevent the expansion with quotes (single or double)

```
args = ["'$(location //multiple_files_target)'"]
```
or
```
args = ["\"$(location //multiple_files_target)\""]
```
or
```
args = ["'my args with spaces'"]
```
or
```
args = ["\"my args with spaces\""]
```
to keep it as a single arg

File: lib/private/strings.bzl (hunk 4 above)
Comment (gregmagolan) [reply lib/private/strings.bzl:587]: It would be invoked once per arg per target at analysis time so it could be an issue if there are lots of targets with args in a repo. In the `jq` rule we added an opt-out from this processing:
```
expand_args: Run bazel's location and make variable expansion on the args.
```
=>
```
    if ctx.attr.expand_args:
        args = []
        for a in ctx.attr.args:
            args += split_args(expand_variables(ctx, ctx.expand_location(a, targets = ctx.attr.data), outs = [out]))
    else:
        args = ctx.attr.args
```
So if needed we can add the same opt-out to other rules in the future

File: lib/private/run_binary.bzl (hunk 3 above)
Comment (jbedard) [reply lib/private/run_binary.bzl:51]: I'm still wondering about this, and maybe just overthinking it 🤔 

Should these all be the exact same?
```
args = ["--one --two --three"]
args = ["--one --two", "--three"]
args = ["--one \"--two\" --three"]
```
🤷

File: lib/strings.bzl [hunk 5]
```diff
@@ -1,7 +1,8 @@
 "Utilities for strings"
 
-load("//lib/private:strings.bzl", _chr = "chr", _hex = "hex", _ord = "ord")
+load("//lib/private:strings.bzl", _chr = "chr", _hex = "hex", _ord = "ord", _split_args = "split_args")
 
 chr = _chr
 ord = _ord
 hex = _hex
+split_args = _split_args
```
Comment (jbedard) [inline lib/strings.bzl:8]: Can we avoid making this public for now? 🤷

File: lib/private/run_binary.bzl (hunk 3 above)
Comment (gregmagolan) [reply lib/private/run_binary.bzl:51]: Yes, the they should all expand to the equivalent of:
```
args = ["--one", "--two", "--three"]
```

File: lib/strings.bzl (hunk 5 above)
Comment (gregmagolan) [reply lib/strings.bzl:8]: It may be be useful to downstream rulesets that are have the same issues. I passed the naming by Alex and he was was with it.

Comment (jbedard) [reply lib/strings.bzl:8]: Lets just be careful when introducing this to other rules 😅

**Review by jbedard — APPROVED** [review]
